feat(react-table): convert react-table to typescript

priley86 opened the pull request:
<!--
Thanks for your interest in patternfly-react. We appreciate all issues filed and PRs submitted!

Please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What issue is being addressed here?) -->

**What**:
Closes #1950 #2504  - React-Table TypeScript conversion.

- [x] Convert `Table` and many of the existing React-Table wrapping components to TSX
- [x] Add temporary `.d.ts` files for some components missing types so that react-table can be compiled w/ `noImplicitAny` set, cc: @ssilvert
- [x] update some faulty types in react-core (such as `InternalDropdownItem`)
- [x] convert `base` Reactabular source to TSX
- [x] reimplement `base` Reactabular usage of Context w/ modern `React.createContext` syntax
- [x] convert react-core `Dropdown` to typescript - we'll need `DropdownPosition` and `DropdownDirection` types in react-core at least
- [x] convert remaining components to typescript
- [x] fix jest tests/convert to typescript

<!-- Are there any upstream issues or separate issues you need to reference? -->

**Additional issues**:

<!-- feel free to add additional comments -->

patternfly-build commented:
PatternFly-React preview: https://patternfly-react-pr-2360.surge.sh

priley86 commented:
cc: @jenny-s51 - please feel free to begin converting tests in `react-table` to typescript if you have time! I'm trying to tackle the remaining components now...

priley86 commented:
cc: @mturley - will definitely try to review this w/ you once through w/ first pass!

mturley commented:
Absolutely @priley86 ! At first glance this looks awesome, but I'll dig a little deeper and get back to you. Maybe early next week we can step through it together on a call.

This is probably outside the scope of this PR, but I notice you're pulling in tslint.. do you have any opinion on moving to eslint-typescript in the future? or at least enabling the prettier plugin for tslint? currently we're not enforcing code formatting in our CI: https://github.com/patternfly/patternfly-react/issues/2362

It might be nice to resolve that either before or shortly after we add a bunch of new code like this, so we don't have as much churn later on.

ssilvert commented:
On 7/19/2019 3:13 PM, Mike Turley wrote:
>
> Absolutely @priley86 <https://github.com/priley86> ! At first glance 
> this looks awesome, but I'll dig a little deeper and get back to you. 
> Maybe early next week we can step through it together on a call.
>
> This is probably outside the scope of this PR, but I notice you're 
> pulling in tslint.. do you have any opinion on moving to 
> eslint-typescript in the future? or at least enabling the prettier 
> plugin for tslint? currently we're not enforcing code formatting in 
> our CI: #2362 <https://github.com/patternfly/patternfly-react/issues/2362>
>
> It might be nice to resolve that either before or shortly after we add 
> a bunch of new code like this, so we don't have as much churn later on.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/patternfly/patternfly-react/pull/2360?email_source=notifications&email_token=AAD32QQZV2IDQPAWAQREBILQAIG5ZA5CNFSM4H3SPW6KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2MQGRQ#issuecomment-513344326>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AAD32QRVYBKHLGWSWR3FKULQAIG5ZANCNFSM4H3SPW6A>.
>
FWIW, Keycloak is using eslint with the @typescript-eslint/eslint-plugin 
and @typescript-eslint/parser.

I chose this because the TypeScript team is embracing eslint over 
tslint: https://eslint.org/blog/2019/01/future-typescript-eslint

priley86 commented:
heh ;) - you guys read my mind, I am actually in full support of this move to `@typescript-eslint/eslint-plugin` and had brought this up in previous emails as well...

I chose to follow `@patternfly/react-core`, but I'm perfectly fine with making the shift off of tslint if others are. I don't if this is something we want to start exploring here. This probably needs further discussion, but it's very important!

cc: @dgutride thoughts on this topic?

priley86 commented:
I realize there was some existing history and this move off of tslint will take some time...

mturley commented:
@priley86 it sounds like it's outside the scope of this PR, but let's at least make sure we run Prettier on this code so it has a good start formatting-wise :) It might be worth getting Prettier to run as part of our tslint usage as a first step before we move to eslint.. I'm less concerned about which linter we have and more about just the fact that we have style rules disabled.

As for the rest of this PR, I'll be mostly buried in meetings the next two days, so I hope to start a more thorough review on Wednesday.

patternfly-build commented:
PatternFly-React preview: https://patternfly-react-pr-2360.surge.sh

priley86 commented:
> @priley86 it sounds like it's outside the scope of this PR

Yea, i believe it should be a different PR, but I'm glad others support this move to `@typescript-eslint/eslint-plugin` too.

> It might be worth getting Prettier to run as part of our tslint usage as a first step before we move to eslint
added tslint and some tslint fixes:
https://github.com/patternfly/patternfly-react/pull/2360/commits/adf97b7faf7f2ace19ee2bf5b3c6585c31c4f1a4

I've disabled these tslint rules since they will currently be breaking changes:
```
    "interface-name": false,
    "no-empty-interface": false,
    "ban-types": [true, {"Function": false}],
```

patternfly-build commented:
PatternFly-React preview: https://patternfly-react-pr-2360.surge.sh

patternfly-build commented:
PatternFly-React preview: https://patternfly-react-pr-2360.surge.sh

mturley commented:
(still probably out of scope, sorry to clutter this thread, but one last thing) I see you added `tslint-config-prettier`, but just fyi that brings us to the same problem we have in react-core. tslint-config-prettier disables tslint rules that would conflict with prettier, but it doesn't run prettier. we should also add `tslint-plugin-prettier` in both places (or set up Prettier CLI to be run as part of CI) https://github.com/patternfly/patternfly-react/issues/2362

patternfly-build commented:
PatternFly-React preview: https://patternfly-react-pr-2360.surge.sh

priley86 commented:
>  we should also add tslint-plugin-prettier in both places
agreed! It does not look like it's functioning correctly. I'll give this a go after other changes are complete!

patternfly-build commented:
PatternFly-React preview: https://patternfly-react-pr-2360.surge.sh

patternfly-build commented:
PatternFly-React preview: https://patternfly-react-pr-2360.surge.sh

priley86 commented:
hey @jenny-s51 ... i've converted all of Table types in the root directory now to TSX. Some things we still need:

* convert utils (transformers and formatters)
* convert RowWrapper.test.js, formatters.test.js, headerUtils.test.js, transformers.test.js
 
Ping me tomorrow if you have a chance and we can split this effort up! Getting closer now on this...

patternfly-build commented:
PatternFly-React preview: https://patternfly-react-pr-2360.surge.sh

priley86 commented:
added cellActions, cellWidth, classNames, collapsible, compoundExpand transforms, refactored `IActions` interfaces to inherit `IActionsItem` and Core `DropdownItem` correctly now.

priley86 commented:
@jenny-s51 i've merged your latest commits in this branch. This is fantastic. Getting soo close here.

priley86 commented:
* squashed 20 commits 
* rebased latest master
* updated `ActionsColumn` to consume the `Dropdown` from react-core (with types)

Removing the W.I.P. label. This is ready for review. 
cc: @dgutride @karelhala @dlabrecq @mturley @redallen @ssilvert 

I'm going to test this downstream in OpenShift Console locally. Will also try to add some Cypress tests during the review...thanks!

priley86 commented:
cc: @suomiy - i think a follow up here should be to convert Inline Edit to typescript. I've eased some of the prop-types extensions we had there for now, but you could import all of the Table types exported here in types.tsx (base Reactabular types) and Table.tsx (react-table extended types)  in your extension.

types.tsx
https://github.com/patternfly/patternfly-react/pull/2360/files#diff-889d6aa319bf7253421151e9158e65b4

Table.tsx
https://github.com/patternfly/patternfly-react/pull/2360/files#diff-9bda7132cc01104382345447864888d5

priley86 commented on packages/patternfly-4/react-table/src/components/Table/utils/transformers.tsx:
```diff
@@ -0,0 +1,40 @@
+  scope: '',
+  component: 'td'
+});
+emptyTD.prototype.name = 'emptyTD';
```
extending function prototype like this is breaking downstream... will find a different workaround. This mainly just to preserve existing tests using the prototype `name` on `Function`, however Typescript does not support this.

priley86 replied:
it seems that some dependency updates on my local have resolved this. Should not be necessary to extend `prototype`.

priley86 commented:
* tested successfully downstream in Console. No build errors or new log errors after the commits above.

priley86 commented:
rebased #2615 (`wrappable` transform merged today) and converted to tsx...

I am deferring cypress/e2e tests for now (this PR is already big enough). Will circle back for this.

Would love to merge this soon if possible... 🎲 🎲 🎲

priley86 commented:
Note: I opted to change `ICell` input types to `any` for the time being. The individual formatter and transformer types can be seen and types are visible downstream now for the imports (so that much is good), however I think we need to shore up those inputs before explicitly typing those props further in `Table.tsx`. This will require more comprehensive testing and further delay this PR, so I think using `any` is OK for now. We can come back and type them more explicitly later I'm thinking.

https://github.com/patternfly/patternfly-react/issues/2504#issuecomment-517370180

```
export interface ICell {
  title?: string;
  transforms?: ((...args: any) => any)[];
  cellTransforms?: ((...args: any) => any)[];
  columnTransforms?: ((...args: any) => any)[];
  formatters?: ((...args: any) => any)[];
  cellFormatters?: ((...args: any) => any)[];
  props?: any;
  data?: any;
  header?: any;
  cell?: any;
}
```

priley86 commented:
* rebased

dgutride commented:
This looks good to me - I noticed a few others have reviewed without approving - @mturley @dlabrecq  - can we get an official approval on this?

karelhala commented on packages/patternfly-4/react-table/src/components/Table/Body.tsx:
```diff
@@ -115,32 +110,23 @@ class ContextBody extends React.Component {
-    {({ headerData, rows }) => <ContextBody headerData={headerData} rows={rows} {...props} />}
+    {({ headerData, rows }) => 
+      <ContextBody 
+        headerData={headerData} 
```
I think that rest of props are missing in here.

priley86 replied:
thanks! added `...rest` for Context props as well (so headerRows will be passed in): https://github.com/patternfly/patternfly-react/pull/2360/commits/fe06c8dc05fff246931db441ad24cf306ff742e9

karelhala commented:
I checked it in Cloud services in inventory table and everything looks and behaves same as before. Nice!

priley86 commented:
* rebased again and addressed above [comment](https://github.com/patternfly/patternfly-react/pull/2360#discussion_r312059672)

Captured #2673 for followup...

Should be g2g!

mturley commented on packages/patternfly-4/react-table/src/components/Table/Table.tsx:
```diff
@@ -0,0 +1,298 @@
+}
+
+export const TableContext = React.createContext({
+  headerData: null as any,
```
I think this `any` could be `RenderersTypes['columns']` based on the fact that `headerData` is passed down to the `columns` prop of `Provider` which has that type.

priley86 replied:
good one! updating.

mturley commented on packages/patternfly-4/react-table/src/components/Table/Table.tsx:
```diff
@@ -0,0 +1,298 @@
+
+    if (!ariaLabel && !caption && !header && role !== 'presentation') {
+      // tslint:disable-next-line:no-console
+      console.error('Table: Specify at least one of: header, caption, aria-label')
```
This is interesting.. I wonder if there is a way to use TypeScript to enforce this.

We might be able to break the props down into separate interfaces where each one has one of the three props required and the other two optional (and another interface with all 3 of those props optional and `role` is specifically only allowed to be `'presentation'`. Then we could `export interface TableProps = TablePropsWithHeader | TablePropsWithCaption | TablePropsWithAriaLabel | TablePropsWithRolePresentation` (all of which extend `BaseTableProps` which doesn't have any of those 4 props). Maybe that's not worth it, but we could avoid having to do this check at runtime.

Inspiration: https://stackoverflow.com/a/40510700

priley86 replied:
interesting... yea i mimicked react-core w/ this approach, but I'm not opposed to that type of change. I think static type checking and the runtime error are BOTH beneficial. Maybe a separate  followup issue or piggybacking on #2673 w/ this for now...

mturley commented on packages/patternfly-4/react-table/src/components/Table/base/body-row.tsx:
```diff
@@ -1,29 +1,40 @@
+  };
+
+  shouldComponentUpdate(nextProps: BodyRowProps){
     const previousProps = this.props;
```
We can probably remove this `previousProps`, right? It was never used anyway, i'm not sure why we had the `// eslint-disable-line no-unused-vars`

priley86 replied:
yes... that is old code in Reactabular ;) - good call... updated to destructure those from `this.props` instead:

```
const {columns, rowData} = this.props;
```

mturley commented on packages/patternfly-4/react-table/src/components/Table/base/body-row.tsx:
```diff
@@ -1,29 +1,40 @@
-        return renderers.row.shouldComponentUpdate.call(this, nextProps);
+    if (renderers && renderers.row && (renderers.row as React.Component).shouldComponentUpdate) {
+      if (isFunction((renderers.row as React.Component).shouldComponentUpdate)) {
+        return (renderers.row as React.Component).shouldComponentUpdate.call(this, nextProps, {}, {});
```
This is really weird and sort of not how `shouldComponentUpdate` is supposed to work, but I'm not going to question it because it was preexisting :)

priley86 replied:
Yes.... yes it is weird.... 

This is all existing logic in Reactabular. FWIW, I don't think calling `shouldComponentUpdate` on a consumer passed component is good React practice, but that is seen as an extendability "feature" from the Reactabular source... 

It also makes typing really hard... I'm OK w/ removing this if others don't see the need to use it. It could help on some edge cases where you are extending body-row though...

priley86 replied:
Probably better ways to write this in the future...

mturley replied:
Let's leave it for now, I'm sure it's black magic that is necessary somehow and we'll probably think of a better way eventually.

mturley commented on packages/patternfly-4/react-table/src/components/Table/base/evaluate-transforms.ts:
```diff
@@ -1,16 +1,17 @@
   if (process.env.NODE_ENV !== 'production') {
     if (!transforms.every(isFunction)) {
-      throw new Error("All transforms weren't functions!", transforms);
+      throw new Error('All transforms weren\'t functions!');
```
Why not display `transforms` as part of this error? Did TS complain there?

priley86 replied:
yes, our version of TypeScript does not allow passing a second param to `ErrorConstructor`... 

![Screen Shot 2019-08-09 at 2 58 12 PM](https://user-images.githubusercontent.com/4237045/62802564-28a18900-bab6-11e9-8629-adc13db2a810.png)

mturley replied:
Ah weird, that makes sense though. 👍

mturley commented on packages/patternfly-4/react-table/src/components/Table/base/merge-props.ts:
```diff
@@ -16,7 +16,7 @@ function mergePropPair(...props) {
-  return mergeWith(mergeWith({}, firstProps), ...restProps, (a, b, key) => {
+  return mergeWith(mergeWith({}, firstProps), ...restProps, (a: any, b: any, key: any) => {
     if (key === 'children') {
       // Children have to be merged in reverse order for Reactabular
```
This is also super weird and maybe worth looking at in the future.. or maybe I just don't understand reactabular well enough :)

priley86 replied:
yea... this is merging cell and cell.props into a single prop object... feel free to ehance in the future!

```
        return React.createElement(
          renderers.cell as createElementType,
          {
            key: `${columnIndex}-cell`,
            ...mergeProps(props, cell && cell.props, transformed)
          },
```

mturley commented on packages/patternfly-4/react-table/src/components/Table/base/provider.tsx:
```diff
@@ -0,0 +1,58 @@
+        header: { ...TableDefaults.renderers.header, ...finalRenderers.header },
+        body: { ...TableDefaults.renderers.body, ...finalRenderers.body }
+      }
+    }}> 
```
More weird spacing that Prettier will fix 😉

mturley commented on packages/patternfly-4/react-table/src/components/Table/utils/decorators/cellActions.tsx:
```diff
@@ -17,24 +18,24 @@ export default (actions, actionResolver, areActionsDisabled) => (
     rowData,
     extraData
-  );
+  ) as boolean;
```
It seems like it's possible for this to be falsy but not `false`, which could throw an unnecessary TypeError, right? Or does `as boolean` actually cast to a boolean? (I need to read up on TS a bit more)

priley86 replied:
again, believe we are just giving static type return types - this won't change the existing logic, only attempt to give static analysis hints.

mturley commented on packages/patternfly-4/react-table/src/components/Table/utils/decorators/classNames.ts:
```diff
@@ -13,6 +13,7 @@ export const Visibility = pickProperties(styles.modifiers, [
 
-export default (...classNames) => () => ({
+// tslint:disable-next-line:no-shadowed-variable
+export const classNames = (...classNames: any[]) => () => ({
```
would `classNames` not be a `string[]`?

priley86 replied:
seems Ok... i was thinking of the case classNames are passed as an object... but it think that's fine for now. Updating!

mturley commented on packages/patternfly-4/react-table/src/components/Table/utils/decorators/collapsible.tsx:
```diff
@@ -34,22 +37,23 @@ export const collapsible = (value, { rowIndex, columnIndex, rowData, column, pro
   ) =>
     rowData.hasOwnProperty('parent') && {
-      colSpan: colSpan + !!rowData.fullWidth,
+      // todo: rewrite this logic, it is not type safe
```
We should probably reference this somewhere so we don't forget about it, maybe in #2673 ?

priley86 replied:
yes, this is also not really Typescript compatible... will add to #2673

mturley commented on packages/patternfly-4/react-table/src/components/Table/utils/decorators/headerCol.tsx:
```diff
@@ -0,0 +1,15 @@
+    const result = typeof value === 'object' ? value.title : value;
+    return {
+      component: 'th',
+      children: <div id={id + rowIndex}>{result}</div>
```
We should probably avoid concatenating strings with `+` right? this could be:
```
<div id={`${id}${rowIndex}`}>
```
Not sure if it matters.

priley86 replied:
updating!

mturley commented on packages/patternfly-4/react-table/src/components/Table/utils/decorators/selectable.tsx:
```diff
@@ -23,9 +24,12 @@ export default (label, { rowIndex, columnIndex, rowData, column, property }) =>
+  function selectClick(event: React.ChangeEvent<HTMLInputElement>) {
     const selected = rowIndex === undefined ? event.target.checked : rowData && !rowData.selected;
-    onSelect && onSelect(event, selected, rowId, rowData, extraData);
+    // todo: change event type to React.FormEvent<HTMLInputElement> in the future, breaking change a.t.m.
```
are we tracking this in an issue?

priley86 replied:
adding to #2673

mturley replied:
Thanks!

mturley commented on packages/patternfly-4/react-table/tsconfig.base.json:
```diff
@@ -0,0 +1,26 @@
+    "skipLibCheck": true,
+    "strict": true,
+    "strictNullChecks": false,
+    // "noImplicitAny": false,
```
`//` is invalid in JSON, does this parse? we should probably either set it or remove it.

priley86 replied:
removed

priley86 commented on packages/patternfly-4/react-inline-edit-extension/src/components/InlineEdit/editableRowWrapper.js:
```diff
@@ -17,7 +16,6 @@ import { inlineEditCss, inlineEditStyles as styles } from './css/inline-edit-css
 inlineEditCss.inject();
 
 const propTypes = {
-  ...ReactTableRowWrapper.propTypes,
```
yes, once we moved `RowWrapper` to `.tsx` I can't reference those `propTypes` the same way. FWIW, these props are the same though. Inline Edit will soon be converted to TypeScript and we can do this w/ interfaces.

mturley replied:
Makes sense! 👍

priley86 replied:
linked in [2679](https://github.com/patternfly/patternfly-react/issues/2679)

priley86 commented on packages/patternfly-4/react-inline-edit-extension/src/components/InlineEdit/editableRowWrapper.js:
```diff
@@ -38,7 +36,6 @@ const propTypes = {
 };
 
 const defaultProps = {
-  ...ReactTableRowWrapper.defaultProps,
```
same as above ^^

mturley replied:
In the meantime before we get this done with interfaces, we're not passing any defaults for those props, though. I just wonder if that could break anything.

priley86 commented on packages/patternfly-4/react-inline-edit-extension/src/components/InlineEdit/editableTableBody.js:
```diff
@@ -118,7 +118,7 @@ const Body = ({ BodyComponent, rows, editConfig, onRowClick, ...props }) => {
 
 Body.propTypes = {
-  BodyComponent: PropTypes.node.isRequired,
+  BodyComponent: PropTypes.any.isRequired,
```
For whatever reason, Jest complains about this now that `Body.tsx` is of type `JSX.Element`. Would like to revisit in InlineEdit rewrite (issue coming) - will reference this.

priley86 replied:
same as above! Will tackle w/ future inline-edit...

priley86 commented on packages/patternfly-4/react-table/src/components/Table/Body.tsx:
```diff
@@ -1,34 +1,29 @@
-  className: PropTypes.string,
-  /** Specify key which should be used for labeling each row. */
-  rowKey: PropTypes.string,
-  /** Function that is fired when user clicks on row.  */
```
Not sure I follow (hope we are comparing the same thing). I am seeing all of the `TableBodyProps` types in the `TableBody`. Even though `rowKey` is not showing correctly, this seems to just be a Gatsby documentation thing... it is of type `RowKeyType` and this should show up in the IDE.

![Screen Shot 2019-08-09 at 2 03 56 PM](https://user-images.githubusercontent.com/4237045/62799342-9053d600-baae-11e9-92e2-d18af7574066.png)

mturley replied:
@priley86 I was just referring to the Description column at the right side of the props table, which is driven by the `/** ... */` comments being removed here.

https://patternfly-react.surge.sh/patternfly-4/components/table/
![Screenshot 2019-08-09 14 30 19](https://user-images.githubusercontent.com/811963/62800881-7b794180-bab2-11e9-9017-b937a6b3173d.png)

priley86 replied:
Hmm yea, I'm not sure. I'm defining defaults in the SFC though...

```
export const TableBody = ({
  onRow = (...args:any) => Object,
  className = '' as string,
  children = null as React.ReactNode,
  rowKey = 'id' as string,
  onRowClick = (...args: any) => undefined as any,
  ...props
}: TableBodyProps) => (
  <TableContext.Consumer>
    {({ headerData = [], rows = [], ...rest }) => 
```

priley86 replied:
ok...seems like the ones passed to `TableBody` are showing up now... marking this resolved for now!

<img width="1038" alt="Screen Shot 2019-08-09 at 6 37 29 PM" src="https://user-images.githubusercontent.com/4237045/62812696-e5a2de00-bad4-11e9-8868-c0a7b898eb66.png">

priley86 commented on packages/patternfly-4/react-table/src/components/Table/Body.tsx:
```diff
@@ -115,32 +110,23 @@ class ContextBody extends React.Component {
-ContextBody.defaultProps = {
-  className: '',
-  children: null,
-  headerData: [],
```
good catch! Updating to default `headerData` and `rows` to default `[]`...the others should be defaulted already.

priley86 commented on packages/patternfly-4/react-table/src/components/Table/BodyWrapper.tsx:
```diff
@@ -0,0 +1,42 @@
+  mappedRows?: IRow[];
+  rows?: IRow[];
+  onCollapse?: (event: React.MouseEvent, rowIndex: number, isOpen: boolean, rowData: IRowData, extraData: IExtraData) => undefined;
+  tbodyRef?: React.Ref<any>;
```
Good point... it looks like TypeScript does not like passing a `Function` to a `tbody`'s ref prop:
```
Type 'Function | ((instance: any) => void) | RefObject<any>' is not assignable to type 'LegacyRef<HTMLTableSectionElement>'.
  Type 'Function' is not assignable to type 'LegacyRef<HTMLTableSectionElement>'.
    Type 'Function' is not assignable to type 'string'.ts(2322)
index.d.ts(97, 9): The expected type comes from property 'ref' which is declared here on type 'DetailedHTMLProps<HTMLAttributes<HTMLTableSectionElement>, HTMLTableSectionElement>'
```

But I do think we should still accept a `Function` for the legacy use case... so for now, adding:

```
tbodyRef?: React.Ref<any> | Function;
```
and casting it beneath:
```
ref={tbodyRef as React.Ref<any>}
```

Let me know if you think of a better option. Not all that familiar w/ this myself!

mturley replied:
That sounds good to me. How does that cast behave if the user passes a function? is that still an error? I'm not too familiar with type assertions yet.

priley86 replied:
Well - i believe it works the same way as passing `ref` prop now in JS. It can be a function or a ref type at runtime. This is just casting the type for static type analysis. Please feel free to double check me on this, however I am seeing that passing a function to `RowWrapper` from our existing `editableRowWrapper` works similarly...and does not give warnings or errors at runtime in Gatsby...

https://github.com/patternfly/patternfly-react/blob/master/packages/patternfly-4/react-inline-edit-extension/src/components/InlineEdit/editableRowWrapper.js#L21

priley86 commented on packages/patternfly-4/react-table/src/components/Table/BodyWrapper.js:
```diff
@@ -1,48 +0,0 @@
-import { mapOpenedRows } from './utils/headerUtils';
-
-// eslint-disable-next-line react/prefer-stateless-function
-class BodyWrapper extends React.Component {
```
Yea, this one was brought up previously I believe. For whatever reason, it should be OK to make this an SFC now.

priley86 commented on packages/patternfly-4/react-table/src/components/Table/CollapseColumn.tsx:
```diff
@@ -0,0 +1,39 @@
+} : CollapseColumnProps ) => {
+  return (
+    <React.Fragment>
+      {isOpen !== undefined && (
```
this is just mimicking the current behavior...
https://github.com/patternfly/patternfly-react/blob/master/packages/patternfly-4/react-table/src/components/Table/CollapseColumn.js#L23

I think it just renders the expand/collapse button on the row if `isOpen` is true or false (if it's undefined, we assume that row does not have it).

mturley replied:
That makes sense. 👍

priley86 commented on packages/patternfly-4/react-table/src/components/Table/Header.tsx:
```diff
@@ -0,0 +1,29 @@
+  ...props
+} : HeaderProps ) => (
+  <TableContext.Consumer>
+  {({ headerRows }) => <ContextHeader {...props} headerRows={headerRows} />}
```
Captured in #2673

mturley replied:
We could run it manually as part of this PR, but I'm ok with waiting until we add it to the build, either way works.

priley86 replied:
yes, would like to keep all the Prettier updates alongside React Core update if OK w/ you.

priley86 commented on packages/patternfly-4/react-table/src/components/Table/RowWrapper.tsx:
```diff
@@ -1,11 +1,35 @@
+}
+
+export interface RowWrapperProps {
+  trRef?: React.Ref<any> | Function;
```
good call! making them consistent.

priley86 commented on packages/patternfly-4/react-table/src/components/Table/RowWrapper.tsx:
```diff
@@ -1,11 +1,35 @@
+  trRef?: React.Ref<any> | Function;
+  className?: string;
+  onScroll?: React.UIEventHandler;
+  onResize?: React.UIEventHandler;
```
I am just trying to map to existing types exposed as closely as possible for now (as to not break existing consumers). Would like to track any future type changes in #2673.

https://github.com/patternfly/patternfly-react/blob/master/packages/patternfly-4/react-table/src/components/Table/RowWrapper.d.ts#L10

This seems to not jive w/ the default `window.addEventListener` callback, so it's being casted to `any` for now.

Will make a note in #2673 !

mturley replied:
Sounds good, sorry if I was nitpicking too many things here that belong in #2673. Let's get this PR merged 😄

priley86 commented on packages/patternfly-4/react-table/src/components/Table/RowWrapper.tsx:
```diff
@@ -64,7 +88,7 @@ class RowWrapper extends React.Component {
       <tr
         {...props}
-        ref={trRef}
+        ref={trRef as React.Ref<any>}
```
It will work and should not complain to the consumer since it eventually resolves to an element. This is just bypassing the static typechecking I guess. FWIW, if all Refs are passed as React.Ref type, TypeScript will not complain, otherwise it seems we'll have to cast to support `Function`.

priley86 commented on packages/patternfly-4/react-table/src/components/Table/RowWrapper.tsx:
```diff
@@ -20,31 +44,31 @@ class RowWrapper extends React.Component {
 
     if (this.props.onScroll) {
-      window.addEventListener('scroll', this.handleScroll);
+      window.addEventListener('scroll', this.handleScroll as any);
```
for whatever reason `Event` and `React.UIEvent<Element>` are not compatible 🙃 

For now, I can remove those `any` casts in the listener, and cast the return type instead...if thats more readable. The current return type is `React.UIEvent<Element>` from our previous definition though.

mturley replied:
You can leave it as-is if you want, I was just curious. Thanks for explaining 😄

priley86 commented on packages/patternfly-4/react-table/src/components/Table/SortColumn.js:
```diff
@@ -1,48 +0,0 @@
-  sortDirection: ''
-};
-
-export const SortByDirection = {
```
just moving the exports here. Tried to gather all of the base types in `base/types.tsx` and all of our extended types in `Table.tsx` so that they are all together for now...

priley86 replied:
lots of types!

priley86 commented:
> Nice work! I pulled the simple example into the seed project and while it seems to render just fine, I get the following runtime warning;
> 
> `Warning: Failed prop type: Invalid prop rows[1] supplied to Table.`
> 
> FWIW I only see this when running in development mode, it isn't reported in production builds.
> 
> I don't see this with the other examples. The only other thing I noticed is that the "Table with headers that wrap" example also uses the `CompactTable` component name, I assume it's just from copy/pasta but could be confusing. Maybe give that example a more relevant name like `WrappableHeadersTable` or something of the sort?

That warning is interesting (and technically correct). Our `SimpleTable` example is using the shorthand to pass the `cells` array without declaring that prop. This technically works b/c Reactabular will still merge those props, however I think we should make all examples consistent w/ our typings. This will probably also lead to more downstream consistency... so I've updated the example code to match typings to resolve this. Consumers still use the shorthand, but I think the warning is appropriate...

Re: "WrappableHeadersTable" - updated!

mturley commented:
@dlabrecq and @karelhala, you guys will need to re-approve, sorry for the churn!

priley86 commented:
thanks for the extensive review @mturley and @seanforyou23 ! Tried to capture all follow-ups in #2673. I believe all breaking changes related to types can come in that future PR... for now, just trying to keep existing consumers unaffected w/ this change.

redallen commented:
@dlabrecq @karelhala If you find anything wrong, please open a follow-up issue!

patternfly-build commented:
Your changes have been released in: 
 - @patternfly/react-inline-edit-extension@2.10.0
 - @patternfly/react-table@2.17.0

Thanks for your contribution! :tada:
